### PR TITLE
Integrate vllm and inference engine (neural speed) 

### PR DESCRIPTION
**Pull request by jiafuzha**

### CODE REVIEW

**Comment by jiafuzha** [comment]
@KepingYan @carsonwang  @xwu99 please help review.

**Comment by xwu99** [comment]
> @KepingYan @carsonwang @xwu99 please help review.

Great work! I will check this week and let you know.

File: vllm-ext/vllm/extension/ns/__init__.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+
+# allow ns quantization
+# ==========================================================================================
+vllm_config = importlib.import_module("vllm.config")
+old_verify_quantization = vllm_config.ModelConfig._verify_quantization
+
+
+def _verify_quntization(self):
+    if self.quantization is not None and self.quantization == "ns":
```
Comment (xwu99) [inline vllm-ext/vllm/extension/ns/__init__.py:37]:
```suggestion
    if self.quantization == "ns":
```

File: vllm-ext/inference_engine/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+# sanitizers
+option(IE_SANITIZE_THREAD        "inference_engine: enable thread sanitizer"                        OFF)
+option(IE_SANITIZE_ADDRESS       "inference_engine: enable address sanitizer"                       OFF)
+option(IE_SANITIZE_UNDEFINED     "inference_engine: enable undefined sanitizer"                     OFF)
+
+# instruction set specific
+option(IE_AVX                    "inference_engine: enable AVX"                                     ON)
+option(IE_AVX2                   "inference_engine: enable AVX2"                                    ON)
+option(IE_F16C                   "inference_engine: enable F16C"                                    ON)
+option(IE_AVX512                 "inference_engine: enable AVX512"                                  OFF)
```
Comment (xwu99) [inline vllm-ext/inference_engine/cpp/CMakeLists.txt:48]: Do we build AVX512 by default? I am not sure the usage of the option here.

File: vllm-ext/vllm/extension/ns/__init__.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    return None
+
+
+vllm_loader._get_quantization_config = _get_quantization_config
+
+# reload to make above changes take effect
+model_loader = importlib.import_module("vllm.model_executor.model_loader")
+importlib.reload(model_loader)
+
+logger.info("__ns extension: use ns model loader for ns model, %s", NSModelLoaderV2.__name__)
```
Comment (xwu99) [inline vllm-ext/vllm/extension/ns/__init__.py:105]: could you summerize a list of which properties are monkey-patched with NS classes in the PR descriptions ?

File: llm_on_ray/inference/predictors/vllm_predictor.py [hunk 4]
```diff
@@ -43,15 +49,44 @@ def __init__(self, infer_conf: InferenceConfig, max_num_seqs):
...
-            device=infer_conf.device,
-            dtype=dtype,
-            disable_log_requests=True,
-            max_num_seqs=max_num_seqs,
-            enforce_eager=infer_conf.vllm.enforce_eager,
-        )
+        if infer_conf.vllm.extension == "ns":
+            logger.warn("applying neural speed extension to vllm ...")
+            try:
+                from vllm.extension import ns as ns
```
Comment (xwu99) [inline llm_on_ray/inference/predictors/vllm_predictor.py:55]:
```suggestion
                from vllm.extension import ns

File: vllm-ext/vllm/extension/ns/model/ns_model.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    ):
+        # get available cores
+        try:
+            max_prompt_tokens = int(os.environ.get(_NS_MAX_PROMPT_TOKENS, "8192"))
+            # cpus_per_work is set to 1 for better ns perf so it's inappropriate to use ray to get available cores
+            # threads = ray.runtime_context.get_runtime_context().get_assigned_resources()['CPU']
+            physical_cores = psutil.cpu_count(logical=False)
+            # reserve one core for non-ns tasks
+            physical_cores = physical_cores if physical_cores <= 1 else physical_cores - 1
+            threads = int(os.environ.get(_NS_NUM_THREADS, str(physical_cores)))
```
Comment (xwu99) [inline vllm-ext/vllm/extension/ns/model/ns_model.py:115]: consider set the cores for ns according to `cpus_per_worker` in the inference config, all other predictors use this config to assign cpu resources.

File: vllm-ext/vllm/extension/ns/model/ns_model.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    def init_inference_engine(
+        self,
+        model_config: ModelConfig,
+        parallel_config: ParallelConfig,
+        scheduler_config: SchedulerConfig,
+    ):
+        # get available cores
+        try:
+            max_prompt_tokens = int(os.environ.get(_NS_MAX_PROMPT_TOKENS, "8192"))
+            # cpus_per_work is set to 1 for better ns perf so it's inappropriate to use ray to get available cores
```
Comment (xwu99) [inline vllm-ext/vllm/extension/ns/model/ns_model.py:110]: you can directly use `cpus_per_work` to set cpu cores for ns, no need to set cpus_per_work to 1 and use another env for ns.

**Comment by xwu99** [comment]
`vllm-ext/inference_engine/cpp/models/application` this directory should be in `vllm-ext/inference_engine/cpp` according to neural_speed code structure

**Comment by jiafuzha** [comment]
> `vllm-ext/inference_engine/cpp/models/application` this directory should be in `vllm-ext/inference_engine/cpp` according to neural_speed code structure

I am not 100% follow neural_speed structure since we only take it as inference engine.

File: vllm-ext/vllm/extension/ns/model/ns_model.py (hunk 6 above)
Comment (jiafuzha) [reply vllm-ext/vllm/extension/ns/model/ns_model.py:110]: no, we cannot. cpus_per_worker value is set to OMP_NUM_THREADS. I actually need to set OMP_NUM_THREADS to 1 to get better perf.

File: vllm-ext/vllm/extension/ns/model/ns_model.py (hunk 5 above)
Comment (jiafuzha) [reply vllm-ext/vllm/extension/ns/model/ns_model.py:115]: It's NS threads which is different from ray cpus. I want to separate them.

File: llm_on_ray/inference/predictors/vllm_predictor.py (hunk 4 above)
Comment (jiafuzha) [reply llm_on_ray/inference/predictors/vllm_predictor.py:55]: good catch

File: vllm-ext/vllm/extension/ns/__init__.py (hunk 1 above)
Comment (jiafuzha) [reply vllm-ext/vllm/extension/ns/__init__.py:37]: good catch

File: vllm-ext/inference_engine/cpp/CMakeLists.txt (hunk 2 above)
Comment (jiafuzha) [reply vllm-ext/inference_engine/cpp/CMakeLists.txt:48]: It's actually not used for now. Instrinsics are checked dynamically during run-time. Will keep it here for later.

File: vllm-ext/vllm/extension/ns/model/ns_model.py (hunk 5 above)
Comment (jiafuzha) [reply vllm-ext/vllm/extension/ns/model/ns_model.py:115]: Just came to my mind that bigger --num-cpus value causes ray starts up more background processes, like client.poll or server.poll processes which hurts overall performance. That's the reason why I set num-cpus to 1 to reduce them. Need to find an elegant way to balance them.

File: vllm-ext/vllm/extension/ns/__init__.py (hunk 3 above)
Comment (jiafuzha) [reply vllm-ext/vllm/extension/ns/__init__.py:105]: All monkey-patches are in this __init__.py. They are,

__init__.py:52] __ns extension: add ns quantization config, NSQuantConfig
__init__.py:105] __ns extension: use ns model loader for ns model, NSModelLoaderV2
__init__.py:116] __ns extension: replace LlamaModel with ns LLamaModel, NSLLamaModel
__init__.py:136] __ns extension: use ns cache engine for ns, NSCPUCacheEngine
__init__.py:146] __ns extension: replace execute_model in cpu_model_runner, execute_model
__init__.py:171] __ns extension: replace BlockSpaceManager.get_block_space_manager_class in vllm.core.interfaces with get_block_space_manager_class

**Comment by jiafuzha** [comment]
It's code before merge. I've addressed all review comments in after merge branch. So, let me close this one and submit a new PR in after merge branch.

**Pull request closed without merging**